audit_log/README.md

```diff
@@ -0,0 +1,209 @@
+| Admin Integration     | Seamlessly integrates with Django admin, automatically logging actions performed through the admin interface, simplifying audit trail management.                                   |
+| Optimized Performance | Includes a specialized paginator for handling large audit log tables, ensuring efficient querying and retrieval of audit data, even with high volumes of events.                    |
+
+## Core Components
+
+- **`AuditLogServiceBase`:**
+
+  - A base class for audit log services, providing common functionality for creating and committing audit log messages.
+  - Decouples audit logging logic from Django's request/response cycle.
+  - Defines methods for determining operation names, extracting actor data, defining targets, and committing messages to loggers and/or the database.
+
+- **`AuditLogApiService`:**
+
+  - Extends `AuditLogServiceBase` to handle Django-specific requests.
+  - Extracts information from Django request objects (e.g., object IDs, operation names, actor data).
+  - Determines if a request endpoint should be audit logged based on configured rules.
+  - Manages the lifecycle of audit log messages within a request (creation, deletion, and committing).
+  - Provides utilities for adding and removing audit logged object IDs from requests.
+
+- **`AuditLogModelAdminMixin`:**
+
+  - A mixin for Django's `ModelAdmin` class to enable audit logging for model admin views.
+  - Automatically logs read, create, update, and delete operations performed through the admin interface.
+  - Integrates with `AuditLogApiService` to commit audit log messages.
+
+- **`LargeTablePaginator`:**
+
+  - A custom paginator optimized for large tables in Django admin.
+  - Uses PostgreSQL's `reltuples` to estimate queryset size for improved performance.
+  - Falls back to the default `Paginator` if `reltuples` is unavailable or unreliable.
+
+- **`AuditLogEntryAdmin`:**
+  - A Django admin class for managing `AuditLogEntry` objects.
+  - Provides a read-only interface for viewing audit log entries.
+  - Uses `LargeTablePaginator` for efficient pagination of potentially large datasets.
+  - Disables creation, modification, and deletion of audit logs through the admin interface.
+  - Displays the audit log message in a user-friendly format.
+
+## Key Features
+
+- **Flexible Logging:** Logs audit events to both a dedicated logger and/or a database based on configuration settings.
+- **Request-Aware:** Extracts relevant information from Django requests to enrich audit log entries.
+- **Configurable:** Allows customization of logged endpoints, request variables, and logging destinations.
+- **Extensible:** Provides a base class for creating custom audit log services tailored to specific needs.
+- **Admin Integration:** Seamlessly integrates with Django admin to audit model operations.
+- **Optimized for Performance:** Includes a specialized paginator for handling large audit log tables.
```

Comment:
This section is somewhat repeating the things that are in the table above, maybe this can be removed?